tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -11499,6 +11503,44 @@ SpirvInstruction *SpirvEmitter::processIntrinsicDP4a(const CallExpr *callExpr,
                                    arg2Instr, loc, range);
 }
 
+SpirvInstruction *SpirvEmitter::processIntrinsicDP2a(const CallExpr *callExpr) {
+  // Processing the `dot2add` intrinsic.
+  // There is no direct substitution for it in SPIR-V, so it is recreated with a
+  // combination of OpDot and OpFAdd.
+  //
+  // float dot2add( half2 a, half2 b, float acc );
+  //    A 2-dimensional floating point dot product of half2 vectors with add.
+  //    Multiplies the elements of the two half-precision float input vectors
+  //    together and sums the results into the 32-bit float accumulator.
+
+  auto loc = callExpr->getExprLoc();
+  auto range = callExpr->getSourceRange();
+
+  assert(callExpr->getNumArgs() == 3u);
+
+  QualType vecType = callExpr->getArg(0)->getType();
+  QualType componentType = {};
+  uint32_t vecSize = {};
+  bool isVec = isVectorType(vecType, &componentType, &vecSize);
+
+  assert(isVec && vecSize == 2);
```

Comment:
Shouldn't both arguments be checked? (or all 3?)

Reply:
This is just an extra assert on things that are already type checked earlier in execution. We don't typically re-check every attribute of every arg here, but if there's anything specific you think might not be caught earlier or should be rechecked I can add it.

Reply:
Actually this needs another fixup for the release build, since isVec is an unused variable. Working on it.

Reply:
Fixed

Reply:
Right. Not a strong opinion, but to me it seems weird to check only one parameter, and trust the other. I'd either check none (assuming some check upstream exists), or check all 3 for the correct type.

Reply:
Ah, this bit of logic is mainly to get the component type, which is used for the dot product. I see how it seems a bit random otherwise.